src/rust/src/buf.rs

```diff
@@ -46,47 +40,130 @@ fn _extract_buffer_length<'p>(
     Ok((bufobj, ptrval))
 }
 
-impl<'a> CffiBuf<'a> {
-    pub(crate) fn from_bytes(py: pyo3::Python<'a>, buf: &'a [u8]) -> Self {
-        CffiBuf {
-            pyobj: py.None().into_bound(py),
-            _bufobj: py.None().into_bound(py),
-            buf,
-        }
+#[cfg(Py_3_11)]
+mod pybuffer_impl {
+    use pyo3::buffer::PyBuffer;
+    use pyo3::types::{PyAnyMethods, PyBytes};
+    use std::slice;
+
+    pub(crate) struct CffiBuf<'p> {
+        pyobj: pyo3::Bound<'p, pyo3::PyAny>,
+        _bufobj: PyBuffer<u8>,
+        buf: &'p [u8],
     }
 
-    pub(crate) fn as_bytes(&self) -> &[u8] {
-        self.buf
+    impl<'a> CffiBuf<'a> {
+        pub(crate) fn from_bytes(py: pyo3::Python<'a>, buf: &'a [u8]) -> Self {
+            let py_bytes = PyBytes::new(py, buf);
+            let pybuffer = PyBuffer::<u8>::get(&py_bytes)
+                .expect("Cannot convert \"bytes\" instance to a buffer.");
+            CffiBuf {
+                pyobj: py.None().into_bound(py),
+                _bufobj: pybuffer,
+                buf,
+            }
+        }
+
+        pub(crate) fn as_bytes(&self) -> &[u8] {
+            self.buf
+        }
+
+        pub(crate) fn into_pyobj(self) -> pyo3::Bound<'a, pyo3::PyAny> {
+            self.pyobj
+        }
     }
 
-    pub(crate) fn into_pyobj(self) -> pyo3::Bound<'a, pyo3::PyAny> {
-        self.pyobj
+    impl<'a> pyo3::conversion::FromPyObject<'a> for CffiBuf<'a> {
+        fn extract_bound(pyobj: &pyo3::Bound<'a, pyo3::PyAny>) -> pyo3::PyResult<Self> {
+            let bufobj = PyBuffer::<u8>::get(pyobj).map_err(|_| {
```

Comment:
Does this reject multi-dimension buffers, if not we should check for and reject them? (I'm 99% sure our existing APIs reject those)

Reply:
Oh, if the existing impl is fine with them, than we shoudl be consistent with that -- I assumed it rejected them.

Reply:
I added
```
            if bufobj.dimensions() != 1 {
                return Err(pyo3::exceptions::PyTypeError::new_err(
                    "Only 1-dimensional byte buffers are supported.",
                ));
            }
```

And the above test passes on the Buffer branch, and fails on the FFI branch so it looks like the test is good.

I'll mark this resolves since the behavior is the same without adding this change

Reply:
Test was run on my fork to avoid eating up CI resources here https://github.com/bdraco/cryptography/pull/7